### PR TITLE
[TS] LPS-177889 - incorrect search results can be returned with certain combinations of delta and results

### PR DESCRIPTION
**Pull request by wanderlast**
Hi Search team!

https://issues.liferay.com/browse/LPS-177889

This is to fix an issue that arises in specific conditions. The following must be true:
- Search results returned must exceed one page (so there should be at least two pages of results).
- The 'leftover' result count should be *over half* of the delta value.

So for example, in the test case presented by the client we have 31 results and a delta of 20. Since 11 results is over 10, this error occurs. Essentially, the error boils down to the fact that:

1. Using the 'from' clause in our query, we already return what is essentially the 'filtered' results for a page.
2. When calling updateDocuments, we attempt to calculate the 'start and end' of the documents we need. However, since we're already filtering it before, we essentially can end up removing results we want to keep. For example, in our test case we're technically looking for results 20-40 (so our start-end) and our 'total/accumulatedCount' returned results is 11. The docs are calculated to therefore start at 20 - 11 or 9 and end at 29. 9 is not greater than our total of '11' so the recalculation that occurs in SearchPaginationUtil never occurs like it would for other values (which is why the leftover result count must be over half of the delta value for this error to occur). 

In fixing this, I made a specific assumption which may be incorrect. I've assumed the start/end calculation are calculations that were based on previously grabbing all results and therefore needing to filter them by refining the start/end values. My solution essentially entails checking if we refined the results using a 'from' clause previously and, if so, skipping the update documents step. I did it in this manner to try to avoid adding any new steps requiring an API but please let me know if there's anything I'm missing or if there's anything you guys would like me to investigate further. Thanks!

### CODE REVIEW

**Comment by joshuacords** [comment]
I'm planning on reviewing this tomorrow.

**Comment by joshuacords** [comment]
Hello @wanderlast, it looks like you found some pretty bad logic here. I'm not exactly sure what you mean by:
> I've assumed the start/end calculation are calculations that were based on previously grabbing all results and therefore needing to filter them by refining the start/end values.
You may mean that when we search on page 2, the start could be set higher that 20 if document were filtered out. This is not true, I tested with a document filtered out and we still are using 20. 
To test I followed the steps on the ticket as described but added a Document Folder with Guest View permission turned off. Then I added a document to that folder named so that it would return early in the search results. (For example: if documents are named "Test Document [X]", I named this "Document".) What I found was in the first page of result, the result number was listed as 30 - but on the second page, the result number is 32!

I looked a little bit deeper and it seems that search being run on the first page goes through the sliding window searcher twice, the first time it searches from 0 to 19, finds and filters out the document inside the permissioned folder, and decrements the counts. Then it searches again, apparently from 1-22(?) pulling much of the same documents including (in my case) the one inside the permissioned folder again. It decrements the count a second time, and then displays the results.

When the search is triggered by navigating to the second page, we search from 20-40 (I believe), in this case the document in the folder is skipped.

In short, this logic for updateDocuments(), standbyDocuments, and the search start and ends is very bad and I'd rather we create some code that addresses these problems directly instead of piping around it. This might be too large of a scope for you to address, but I would appreciate if you could take the first shot at it.

This is the behavior I think we want to see:

1. When searching pages beyond the first, we still search starting at 0 so we filter out any permissioned documents.

**Comment by joshuacords** [comment]
One other thing, [LPS-173591](https://issues.liferay.com/browse/LPS-173591) made some changes to the DefaultSearchResultPermissionFilter that might address at least one of the above issues. You should rebase on top of that code before further investigations.

**Comment by wanderlast** [comment]
Hi @joshuacords,

I appreciate the guidance! I'll take an initial look and hopefully can figure something out. Thanks!
